tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank3TestHelpers.hpp

```diff
@@ -0,0 +1,444 @@
+/// used, and `TensorIndexC` is the third index used..
+///
+/// \tparam Datatype the type of data being stored in the Tensors
+/// \tparam TensorIndexTypeA the \ref SpacetimeIndex "TensorIndexType" of the
+/// first index of the RHS Tensor
+/// \tparam TensorIndexTypeB the \ref SpacetimeIndex "TensorIndexType" of the
+/// second index of the RHS Tensor
+/// \tparam TensorIndexTypeC the \ref SpacetimeIndex "TensorIndexType" of the
+/// third index of the RHS Tensor
+/// \tparam TensorIndexA the type of the first TensorIndex used on the RHS of
+/// the TensorExpression, e.g. `ti_a_t`
+/// \tparam TensorIndexB the type of the second TensorIndex used on the RHS of
+/// the TensorExpression, e.g. `ti_B_t`
+/// \tparam TensorIndexC the type of the third TensorIndex used on the RHS of
+/// the TensorExpression, e.g. `ti_c_t`
+/// \tparam ValenceA the valence of the first index used on the RHS of the
+/// TensorExpression
+/// \tparam ValenceB the valence of the second index used on the RHS of the
+/// TensorExpression
+/// \tparam ValenceC the valence of the third index used on the RHS of the
+/// TensorExpression
+/// \param tensorindex_a the first TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_a`
+/// \param tensorindex_b the second TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_B`
+/// \param tensorindex_c the third TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_c`
+template <typename Datatype,
+          template <size_t, UpLo, typename> class TensorIndexTypeA,
+          template <size_t, UpLo, typename> class TensorIndexTypeB,
+          template <size_t, UpLo, typename> class TensorIndexTypeC,
+          UpLo ValenceA, UpLo ValenceB, UpLo ValenceC,
+          typename TensorIndexA, typename TensorIndexB, typename TensorIndexC>
+void test_evaluate_rank_3_ac_symmetry(const TensorIndexA& tensorindex_a,
+                                      const TensorIndexB& tensorindex_b,
+                                      const TensorIndexC& tensorindex_c) {
+#define DIM_AC(data) BOOST_PP_TUPLE_ELEM(0, data)
+#define DIM_B(data) BOOST_PP_TUPLE_ELEM(1, data)
+#define FRAME(data) BOOST_PP_TUPLE_ELEM(2, data)
+
+#define CALL_TEST_EVALUATE_RANK_3_CORE(_, data)                           \
+  test_evaluate_rank_3_core<                                              \
+      Datatype, Symmetry<2, 1, 2>,                                        \
+      index_list<TensorIndexTypeA<DIM_AC(data), ValenceA, FRAME(data)>,   \
+                 TensorIndexTypeB<DIM_B(data), ValenceB, FRAME(data)>,    \
+                 TensorIndexTypeC<DIM_AC(data), ValenceC, FRAME(data)>>>( \
+      tensorindex_a, tensorindex_b, tensorindex_c);
+
+  GENERATE_INSTANTIATIONS(CALL_TEST_EVALUATE_RANK_3_CORE, (1, 2, 3),
+                          (1, 2, 3), (Frame::Grid, Frame::Inertial))
+
+#undef CALL_TEST_EVALUATE_RANK_3_CORE
+#undef FRAME
+#undef DIM_B
+#undef DIM_AC
+}
+
+/// \ingroup TestingFrameworkGroup
```

Comment:
I would say you don't need to document all these different slightly different variations of this function in such detail. There are a couple options
- Only use `\brief`
- Use Doxygens `// @{` and `// @}` to apply the same comment to all the functions without having to copy-paste or write it out again
- Use Doxygen's `copydoc` (I think that's the right now) function to just copy the doc automatically between the different functions.

Reply:
@nilsdeppe If I remove the `\details` from `test_evaluate_rank_3_no_symmetry`, `test_evaluate_rank_3_ab_symmetry`, `test_evaluate_rank_3_ac_symmetry`, `test_evaluate_rank_3_bc_symmetry`, and `test_evaluate_rank_3_abc_symmetry` I think the the issue still remains that I couldn't use `copydoc` to wholesale copy the documentation around because the `\brief` between them differs. The `\brief` descriptions differ only slightly, but I think the way they differ is what's important to point out - that they are testing different symmetries - because otherwise it might not be clear why there are multiple functions defined that look almost the same. That's just my take though! Honestly, seeing the difference in the names of the functions, alone, might be enough to extract that information. Meaning, maybe I could just have something like this be the shared `\brief` between all and not mention anything about symmetry at all:
```
/// \brief Iterate testing of evaluating single rank 3 Tensors on multiple Frame
/// types and dimension combinations
```
 If you think they can all have the same `\brief` and call it a day, then I think I can simply use `copydoc` and call it a day, because the `\details`, `\tparam`s, and `param`s are what are the same between all of these - only their `\briefs` differ. I saw that there is a way to copy over just the `\details` from another function, but I don't see a way to copy over just the `\tparam`s and `\param`s. Ideally, it would be nice if I could leave each of the `\brief`s and then copy over the `\details` with `\copydetails` and the `\tparam`s and `\param`s with a similar function. Do you know if it's possible to do something like this? The only possible thing I have found online would be to define a separate file with the text to copy over.

So I guess the tldr is: which do you think is better and feasible?
1) changing the `\brief`s of those 5 to be exactly the same (no mention of symmetry, like in the example `\brief` I gave above)
2) keeping the `\brief`s different, but copying the `\details`, `tparam`s, and `param`s (if that's even possible)

Reply:
So what we've done in other situations like this where basically several functions are nearly identical with possibly long comments we have a single comment along the lines of:
```cpp
/// \brief Iterate testing of evaluating single rank 3 Tensors on multiple Frame
/// types and dimension combinations
///
/// We test various different symmetries to ensure that the code works correctly in with
/// symmetries. The symmetries we test are:
/// - A
/// - B
/// - C
```
You can also list other things that are checked, such as different dimensions, and/or valence, and/or spacetime vs spatial index.

Regarding the `\tparam` and `\param`. What we've done here is basically give a list of the parameters without using these keywords. Since Doxygen checks the keywords it becomes a bit tedious to handle various permutations. I would also say you do something like

```
Check with indices A and B that have `ValenceA` and `ValanceB`, `TensorIndexA` and `TensorIndexB` (either spatial or spacetime)
```
Having said that, my view in general is that deduced template parameters don't really need documenting (they are deduced based off the value passed in), while explicit template parameters should be documented (eg DataType).

Having said all this: what if you eliminate the documentation for the `core` functions (they are really the interface and could/should go into a detail namespace), and then trim the documentation for eg `test_evaluate_rank_2_symmetric` to:
```cpp
/// \ingroup TestingFrameworkGroup
/// \brief Iterate testing of evaluating a single rank 2 Tensors on multiple
/// Frame types and dimension combinations for symmetric indices
///
/// \details `TensorIndexA` and `TensorIndexB` can be any type of TensorIndex
/// and are not necessarily `ti_a_t` and `ti_b_t`. The "A" and "B" suffixes just
/// denote the ordering of the generic indices of the RHS tensor expression. In
/// the RHS tensor expression, it means `TensorIndexA` is the first index used
/// and `TensorIndexB` is the second index used.
```
I think that gives all the details you add with the `\tparam` and `\param` specifications. What do you think?

Reply:
@nilsdeppe Thanks so much for your input! I have some comments and questions that became a bit lengthy, so please ping me on slack if it's easier to discuss the things below over slack. Here are my comments:

1) Regarding this suggestion you made:

```
/// \brief Iterate testing of evaluating single rank 3 Tensors on multiple Frame
/// types and dimension combinations
///
/// We test various different symmetries to ensure that the code works correctly in with
/// symmetries. The symmetries we test are:
/// - A
/// - B
/// - C
```

I want to make sure I understand correctly that you mean for this to be something written once and copied to the other functions for different symmetries (e.g. write it once as the documentation for `test_evaluate_rank_3_no_symmetry`, then use `\copydoc` to copy it to the documentation for `test_evaluate_rank_3_ab_symmetry`, `test_evaluate_rank_3_ac_symmetry`, `test_evaluate_rank_3_bc_symmetry`, and `test_evaluate_rank_3_abc_symmetry`). If this is the case:
- I think it reads a little strangely to have "we test different symmetries, here is the list" as part of the documentation within each of these instead of that being stated at a more outer / upper level of documentation for something that contains these functions, because when it's read, it almost sounds like it means "this function tests the following symmetries," but it only tests one of them. But I'm also not strongly opposed to this suggestion if this is the best way to keep things clean and concise. Plus, I can try to think of a way to reword it so it's clear that although we're giving a list of symmetries, that the current function only tests one of them.
- If we were to do this, this solves the issue of repeated documentation for the functions that do the same thing for different symmetries. That group of functions currently has the exact same documentation for the `\details`s, `\param`s, and `\tparam`s, but their `\brief`s differ only slightly in wording due to them dealing with different symmetries. Therefore, it's my opinion that if we do this for something like these 5 functions, I can just write something similar to what this snippet above is and copy it over to the other four functions, but still have that documentation have the `\tparam`s and `\param`s (plus deduced `\tparam`s removed) because those are the same across all 5 functions. This is essentially the option (1) that I mentioned in my previous comment, but with your comments above on testing different symmetries.
- If I'm misunderstanding this suggestion, let me know. Or if you can point me to an example of where we do this elsewhere in the code, let me know

2) Regarding this suggestion you made:

```
/// \ingroup TestingFrameworkGroup
/// \brief Iterate testing of evaluating a single rank 2 Tensors on multiple
/// Frame types and dimension combinations for symmetric indices
///
/// \details `TensorIndexA` and `TensorIndexB` can be any type of TensorIndex
/// and are not necessarily `ti_a_t` and `ti_b_t`. The "A" and "B" suffixes just
/// denote the ordering of the generic indices of the RHS tensor expression. In
/// the RHS tensor expression, it means `TensorIndexA` is the first index used
/// and `TensorIndexB` is the second index used.
```

- This suggestion makes me think I'm misunderstanding what is being called redundant. I had thought that you meant the documentation between something like the following was redundant: `test_evaluate_rank_3_no_symmetry`, `test_evaluate_rank_3_ab_symmetry`, `test_evaluate_rank_3_ac_symmetry`, `test_evaluate_rank_3_bc_symmetry`, and `test_evaluate_rank_3_abc_symmetry`. If this is the case, then I think the problem I mentioned in my previous comment remains - that I can't use something like `\copydoc` to copy around documentation like this between these 5 functions because the `\brief` above is unique to the function due to it specifying which symmetry it deals with (and thus can't be copied around). If I have a misunderstanding of what is being called redundant, let me know!
- I of course will go with whatever is requested of me by you and the other developers, but I personally don't think we should eliminate the documentation for the `core` functions altogether because although they contain the innermost details of the testing, I think it's worth pointing out what these are doing because callers like `test_evaluate_rank_3_ac_symmetry` or its documentation don't capture what `test_evaluate_rank_3_core` is doing. The callers of these `core` functions merely generate a multitude of tensors as test "input," but I think it's worth explaining in the documentation for the `core` functions that these `core` functions are checking that the data is correct when these "input" tensors are evaluated with different LHS tensor index orders, which (in my opinion) is probably the most important detail to mention about the unit test. If you think there is a place to mention this detail that is better or perhaps more consistent with what is done elsewhere in SpECTRE, let me know! I just think it's important to mention that the data is being checked when evaluating with all the LHS index combinations. To me, that place makes sense in the documentation for the `core` functions, but maybe there is a better spot for it!

Thanks again for the suggestions and input!

Reply:
1. I personally like the style in your example. I agree that the wording feels a bit awkward since the function does only one of the symmetries. What about something like "This function tests one of the following symmetries:.... Other functions test the additional symmetries"?
2. That's fair. So something for you to decide then is: how much should be described is doxygen documentation vs. should just be a C++ comment at the start of the or interspersed throughout the function definition. I think this is a matter of taste mostly, so decide something and let's go with that :)

Reply:
1. Done!
2. I went ahead and left a decent amount of documentation. If you think the inline C++ comments in the `core` functions are cluttering or too redundant with their `\details`, let me know and I can remove one of them. Also, if you think this part shouldn't be in both the `core` functions and their callers, let me know which one you think I should remove:
 
```
/// \details `TensorIndexA` and `TensorIndexB` can be any type of TensorIndex
/// and are not necessarily `ti_a_t` and `ti_b_t`. The "A" and "B" suffixes just
/// denote the ordering of the generic indices of the RHS tensor expression. In
/// the RHS tensor expression, it means `TensorIndexA` is the first index used
/// and `TensorIndexB` is the second index used.
```

I left the `\details`, `\tparam`s, and `\param`s for the `core` functions as-is